packages/mdc-density/_functions.scss

```diff
@@ -0,0 +1,70 @@
+//
+// Copyright 2019 Google Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+@import "./variables";
+
+///
+/// Returns component height or size for given density scale and density map.
+///
+/// @param {Number | String} $density-scale - Density scale value for component. Examples: `-3`, `0` or `compact`.
+/// @param {Map} $density-scale-map - Density scale map for the target component.
+///
+/// @example
+///   mdc-density-prop-value(compact, (
+///     default: 36px,
+///     comfortable: 32px,
+///     compact: 24px,
+///   ));
+///   // 24px
+///
+/// @example
+///   mdc-density-prop-value(-2, (
+///     default: 36px,
+///     comfortable: 32px,
+///     compact: 24px,
+///   ));
+///   // 28px
+///
+@function mdc-density-prop-value($density-scale, $density-scale-map) {
+  @if (type-of($density-scale) == "string" and
+  index($list: $mdc-density-supported-scales, $value: $density-scale) == null) {
+    @error "mdc-density: Supported density scales #{$mdc-density-supported-scales}, but received #{$density-scale}.";
+  }
+
+  $height: null;
```

Comment:
This is similar to what I was proposing in #4944, but loses the level of indirection for the property being used. What I was suggesting was that each component has a config that captures more than one property, e.g. both `height` and `width`. So something like
```scss
$mdc-button-density (
  height: (
    0: $mdc-button-height, // Original height prior to density system
    -1: 32px,
    -3: $mdc-button-minimum-height
  ),
  width: (
    0: 10px,
    -1: 20px,
    -3: 30px
  )
  interval: 4px
)
```

Then you could apply the values like
```scss
.whatever-class {
  height:  mdc-density-prop-value($mdc-button-density, compact, height);
  width:  mdc-density-prop-value($mdc-button-density, compact, width);
}
```

I imagine different components would have different properties with specific sizes at different levels. Including the interval in the component config also keeps the door open to custom intervals for each component, even if it does end up being `4px` universally. 

Did you have any reasons why this system would be problematic?

Reply:
I like the idea of density config where we capture config for more than one property. I made changes in this PR accordingly as per your first point.

Regarding your 2nd point, We decided not to change density interval per component, it is always universal (as of now). This universal interval can be overridden by clients. We can always add interval to config if needed.

Also, related to your numbered scales as keys here are the issues that I see with this:

- We also need mapping for token density scale (e.g., `compact`)

So, other alternative config structure would be:

```scss
$density-config: (
  height: (
    default: $mdc-button-height,
    interval: 4px,
  ),
  scale-map: (
    compact: -3,
    comfortable: -1,
  ),
)
```

But this makes it hard for baseline or clients to configure these values relative to baseline. `24px` or `compact` is much more straight forward than numbered scales, IMO. It gets even more complication when we start introducing more properties.

- If we know the default property value we can programmatically determine other numbered scales based on universal interval, so I see that it is redundant.

Reply:
My thinking is that the "compact", "comfortable", and "cozy" names are great for end-users picking from a menu, but that they're completely opaque to a developer. You the components have a fixed interval for density steps, it's much easier to visualize those intervals as "0, -1, -2" than remember which C-word is which. 

I had also been assuming that each of "compact", "comfortable", and "cozy" maps to a specific step universally. Is that not the case?

Reply:
"compact", "comfortable" values do not universally map to same scale. Also, it is not practical to apply same numbered scale "0, -1, -2" across components. For example, the minimum density for button is `-3` whereas the minimum density for checkbox is `-4`. Tokens helps map to appropriate density based on component which makes it easier for developer to use these without hardcoding the numbered scales. i.e., `compact` always refers to lowest supported density scales across components.

Reply:
I suppose that works for me